beetsplug/subsonicplaylist.py

```diff
@@ -0,0 +1,140 @@
+            alt_error = 'error getting playlist, but no error message found'
+            self._log.warn(playlist.attrib.get('message', alt_error))
+            return
+
+        name = playlist.attrib.get('name', 'undefined')
+        tracks = [(t.attrib['artist'], t.attrib['album'], t.attrib['title'])
+                  for t in playlist]
+        return name, tracks
+
+    def commands(self):
+        def build_playlist(lib, opts, args):
+            self._parse_opts(opts)
+            ids = self.config['playlist_ids'].as_str_seq()
+            if len(self.config['playlist_names'].as_str_seq()) > 0:
+                playlists = ET.fromstring(self.send('getPlaylists').text)[
+                    0]
+                if playlists.attrib.get('code', '200') != '200':
+                    alt_error = 'error getting playlists,' \
+                                ' but no erro message found'
+                    self._log.warn(
+                        playlists.attrib.get('message', alt_error))
+                    return
+                for name in self.config['playlist_names'].as_str_seq():
+                    for playlist in playlists:
+                        if name == playlist.attrib['name']:
+                            ids.append(playlist.attrib['id'])
+            playlist_dict = dict()
+            for playlist_id in ids:
+                name, tracks = self.get_playlist(playlist_id)
+                for track in tracks:
+                    if track not in playlist_dict:
+                        playlist_dict[track] = ';'
+                    playlist_dict[track] += name + ';'
+            # delete old tags
+            if self.config['delete']:
+                for item in lib.items('subsonic_playlist:";"'):
+                    item.update({'subsonic_playlist': ''})
+                    with lib.transaction():
+                        item.try_sync(write=True, move=False)
+
+            self.update_tags(playlist_dict, lib)
+            ## notify plugins
+            plugins.send('database_change', lib=lib, model=self)
+
+        subsonicplaylist_cmds = Subcommand(
+            'subsonicplaylist', help=u'import a subsonic playlist'
+        )
+        subsonicplaylist_cmds.parser.add_option(
+            u'-d',
+            u'--delete',
+            action='store_true',
+            help=u'delete tag from items not in any playlist anymore',
+        )
+        subsonicplaylist_cmds.func = build_playlist
+        return [subsonicplaylist_cmds]
+
+    def _parse_opts(self, opts):
+
+        if opts.delete:
+            self.config['delete'].set(True)
+
+        self.opts = opts
+        return True
+
+    def generate_token(self):
+        salt = ''.join(random.choices(string.ascii_lowercase + string.digits))
+        return md5(
+            (self.config['password'].get() + salt).encode()).hexdigest(), salt
+
+    def send(self, endpoint, params=''):
+        url = '{}/rest/{}?u={}&t={}&s={}&v=1.12.0&c=beets'.format(
```

Comment:
It might be best to use `urlencode` to avoid unpleasantness when there are special characters in the parameters:
https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode

Reply:
I was sorta recommending something like this:

```
params = {
    'u': self.config['base_url'].get(),
    't': endpoint,
    ...
}
```

As it stands, doing `+ urlencode(params)` might erroneously concatenate two query strings. (But `params` seems to be unused?)